pandas/core/arrays/period.py

```diff
@@ -82,7 +94,186 @@ def wrapper(self, other):
+            if freq is None or freq == data.freq:  # no freq change
+                freq = data.freq
+                data = data._ndarray_values
+            else:
+                base1, _ = _gfc(data.freq)
+                base2, _ = _gfc(freq)
+                data = libperiod.period_asfreq_arr(data._ndarray_values,
+                                                   base1, base2, 1)
+            return cls._simple_new(data, freq=freq)
+
+        # not array / index
+        if not isinstance(data, (np.ndarray, PeriodIndex,
+                                 DatetimeIndex, Int64Index)):
+            if is_scalar(data) or isinstance(data, Period):
+                # XXX
+                cls._scalar_data_error(data)
+
+            # other iterable of some kind
+            if not isinstance(data, (list, tuple)):
+                data = list(data)
+
+            data = np.asarray(data)
+
+        # datetime other than period
+        if is_datetime64_dtype(data.dtype):
+            data = dt64arr_to_periodarr(data, freq, tz)
+            return cls._from_ordinals(data, freq=freq)
+
+        # check not floats
+        if lib.infer_dtype(data) == 'floating' and len(data) > 0:
+            raise TypeError("PeriodIndex does not allow "
+                            "floating point in construction")
+
+        # anything else, likely an array of strings or periods
+        data = ensure_object(data)
+        freq = freq or libperiod.extract_freq(data)
+        data = libperiod.extract_ordinals(data, freq)
+        return cls._from_ordinals(data, freq=freq)
+
+    @property
+    def asi8(self):
+        return self._data.view("i8")
+
+    @classmethod
+    def _from_sequence(cls, scalars, dtype=None, copy=False):
+        return cls(scalars, dtype=dtype, copy=copy)
+
+    @classmethod
+    def _from_factorized(cls, values, original):
+        return cls(values, dtype=original.dtype)
+
+    def __repr__(self):
+        return '<pandas PeriodArray>\n{}\nLength: {}, dtype: {}'.format(
+            [str(s) for s in self],
+            len(self),
+            self.dtype
+        )
+
+    def __len__(self):
+        return len(self._data)
+
+    def isna(self):
+        return self._data == iNaT
+
+    def take(self, indices, allow_fill=False, fill_value=None):
+        from pandas.core.algorithms import take
+
+        if fill_value is None:
+            fill_value = iNaT
+        elif isinstance(fill_value, Period):
+            fill_value = fill_value.ordinal
+        elif fill_value is NaT:
+            fill_value = iNaT
+        elif fill_value != self.dtype.na_value:
+            raise ValueError("Expected a Period.")
+
+        new_values = take(self._data,
+                          indices,
+                          allow_fill=allow_fill,
+                          fill_value=fill_value)
+
+        return self._from_ordinals(new_values, self.freq)
+
+    @property
+    def nbytes(self):
+        return self._data.nbytes
```

Comment:
Off topic: this kind of thing could definitely go in a core.arrays.common mixin shared across pandas-internals EAs

Reply:
Not all of the arrays will support this (but maybe enough do to share it ...), eg IntervalArray has two arrays, Categorical in principle has also both codes and categories

(in any case, I would also leave that for later, if we have the multiple arrays, we can see which functionality is common and can be moved to a mixin/superclass, but now with the different arrays in different PRs this will be a bit difficult)

Reply:
Will postpone (though certainly all of DatetimeArray, TimedeltaArray, and PeriodArray will share this).